File: LeetCode/Explore/TopInterview/Blind75/Array/121_best_time_to_buy_and_sell_stock.cpp
// one pass
// Complexity Analysis
// Time complexity: O(n)O(n). Only a single pass is needed.
// Space complexity: O(1)O(1). Only two variables are used.
  
class Solution {
public:
  int maxProfit(vector<int>& prices) {
    int maxProfit = 0, minPurchase = prices[0];
    for (int i = 0; i < prices.size(); i++) {
      if (prices[i] < minPurchase) {
        minPurchase = prices[i];
      }
      else {
        maxProfit = max(maxProfit, prices[i] - minPurchase);
      }
    }
    return maxProfit;
  }
};

File: LeetCode/Explore/TopInterview/Blind75/Array/121_best_time_to_buy_and_sell_stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
      minCost = prices[0]
      maxGain = 0
      
      for i, price in enumerate(prices):
        if price < minCost:
          minCost = min(minCost, price)
        elif price > minCost:
          maxGain = max(maxGain, price - minCost)
        else:
          continue
          
      return maxGain
